crates/aptos-crypto/src/bls12381/bls12381_sigs.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides APIs for aggregating and verifying BLS multi-signatures
+//! implemented on top of BLS12-381 elliptic curves (https://github.com/supranational/blst).
+//!
+//! The `BLS12381Signature` struct is used to represent either a signature share from an individual
+//! signer or a multisignature aggregated from many such signers.
+//!
+//! The signature verification APIs in `Signature::verify` and `Signature::verify_arbitrary_msg` do NOT
+//! assume the signature to be a valid group element and will implicitly "group-check" it. This
+//! makes the caller's job easier and, more importantly, makes the library safer to use.
+
+use std::convert::TryFrom;
+use anyhow::{anyhow, Result};
+use blst::BLST_ERROR;
+use serde::Serialize;
+use aptos_crypto_derive::{DeserializeKey, SerializeKey};
+use crate::bls12381::bls12381_keys::{BLS12381PublicKey, BLS12381PrivateKey};
+use crate::hash::CryptoHash;
+use crate::{CryptoMaterialError, Length, Signature, signing_message, ValidCryptoMaterial, ValidCryptoMaterialStringExt};
+use crate::bls12381::DST_BLS_MULTISIG_IN_G2_WITH_POP;
+
+#[derive(Debug, Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// Either A BLS signature share from an individual signer or a BLS multisignature aggregate from
+/// multiple such signers
+pub struct BLS12381Signature {
+    pub(crate) sig: blst::min_pk::Signature,
+}
+
+////////////////////////////////////////////////
+// Implementation of (multi)signature structs //
+////////////////////////////////////////////////
+
+impl BLS12381Signature {
+    /// The length of the BLS12381Signature
+    // NOTE(Alin): We have to hardcode this here because there is no library-defined constant
+    pub const LENGTH: usize = 96;
+
+    /// Serialize a BLS12381Signature.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.sig.to_bytes()
+    }
+
+    /// Group-checks the signature: i.e., verifies the signature is a valid group element.
+    /// WARNING: This is called implicitly when verifying the signature in this struct's
+    /// `Signature::verify_arbitrary_msg` trait implementation. Therefore, this function should not
+    /// be called separately for most use-cases. We leave it here just in case.
+    pub fn validate(&self) -> Result<()> {
+        self.sig.validate(true).map_err(|e| anyhow!("{:?}", e))
+    }
+
+    /// Aggregate multiple signatures; does NOT assume individual signature are validated NOR group-checked.
+    /// For performance reasons, this function will not group-check the signature shares. Since the
+    /// verification of the returned multisignature will include such a group check, we are not
+    /// concerned by adversarial signature shares.
```

Comment:
probably shouldn't say we're not concerned by adversarial signature shares. it's more like we aggregate optimistically and will fall back to individual checks if multisig verification fails